docs/hosting/config.md
path: tree/master/data
source: config_template.yml

# Configuration
Some aspects of NabBot can be customized using its `config.yml` file.
This file is generated on the first startup, based on the template found in `data/config_template.yml`.
On every startup, the config file is checked and it notifies via console if any key is missing or there's a key not supported.

If a key is missing, a default value is taken, so the bot is still able to function without trouble.

If there's an extra key found, this can mean that a previous configuration entry was removed or a typo was made.

The correct structure can always be found in `config_template.yml`.

## Channels
```yaml
ask_channel_name: ask-nabbot
ask_channel_delete: true
log_channel_name: server-log
```

Specifies the names of NabBot's special channels. The ask channel is a special channel where NabBot will give longer replies.
This is used to reduce spam in main channels by keeping responses short.

The `ask_channel_delete` key specifies if messages that are not commands in the ask channel should be deleted or not.
This way you can keep #ask-nabbot as a commands only channel.

The log channel is a special channel where NabBot posts server events such as members joining, leaving or getting banned.
The bot will announce the following events here:

- A new member joined, with a list of their previously registered characters, if available
- A member left or was kicked (if the bot has `View Audit Log` permission, it can tell the difference)
- A member was banned
- A member changed his name or nick.
- A member registered or unregistered characters.
- The server's name was changed
- The server's region was changed.

## Command prefix
```yaml
command_prefix:
  - "/"
command_mention: false
```

The prefix for commands that the bot will use. By default, the bot will listen to commands like: `/about`, `/help`.
The prefix `/` can be changed, or even more prefixes can be added like:

```yaml
command_prefix:
  - "/"
  - "$"
```

So the bot would now also answer to `$about` and `$help`.  
Note that the bot will always answering by being mentioned, e.g. `@NabBot help`, `@NabBot about`.

It's recommended to keep this list as short as possible, and to make sure it does not overlap with the command prefix of other bots.

This setting can be overriden on a per-server basis by using the command [settings](../commands/admin.md#settings-prefix)
    
## Extra cogs
For more information, see [Cogs](cogs.md)

```yaml
# Add extra features by adding your own cogs.
extra_cogs: []
```

## Owner IDs
```yaml
owner_ids:
  - 162060569803751424
  - 162070610556616705
```

This gives the users with those user ids permission to use any commands and by pass most regulations.
The owner of the bot's application is always considered even if their id is not here.

This allows them to use sensitive commands like `/shutdown` and `/restart` or execute Python code directly.

## Announce Threshold
```yaml
announce_threshold: 30
```

This is the mininum level for NabBot to announce levels and deaths. Note that even if they are not announced, they are still tracked and stored.

Checking a character directly using `/deaths` or `/levels` will show all entries, but seeing them in overall lists using the commands without parameters will hide such entries.

## Online List Expiration
```yaml
online_list_expiration: 300
```

In order to prevent losing level up announcements because NabBot was restarted, the state of online players is saved in a file.
However, if the data is too old, it must be discarded to prevent errors.

This is in the interval in seconds to consider the online list still valid.

## Scan intervals
```yaml
# Delay inbreed server checks
online_scan_interval: 40

# Delay in between player death checks in seconds
death_scan_interval: 15

# Delay between retries when there's a network error in seconds
network_retry_delay: 1
```

These are intervals related to fetching operations.
These were relevant when Tibia.com was used for most of the data, to reduce errors due to CipSoft blocking constant requests.

Now that TibiaData is used, this is not as relevant, as they use caching.

This might be removed in future updates.

## Emojis
Some information is displayed using emojis, to make it easier to identify at quick glance.
These emojis can be personalized by editing the configuration file.

Apart of unicode (standard) emojis, custom discord emojis can be used.
Discord bots are able to use emojis from any server they are in, anywhere, like they had [Nitro](https://discordapp.com/nitro).

It is recommended to use custom emojis from a server where no one else can modify them, as this may break NabBot.
A server dedicated for emojis may be created, inviting the bot for it to use them.

To use a custom emoji, you have to declare it like: `<:name:id>`, for example: `<:fireDamage:458794525050142740>`.
For animated emojis, you have to add `a`, example: `<a:paladin:45233654759231>`.

To find out the id of an emoji, there's two ways to do it:

1. Right click and emoji, select `Copy Link`, a link like `https://cdn.discordapp.com/emojis/458794525050142740.png?v=1` will be created.  
   The numeric part is the emoji's id.
2. Send a message with the emoji, placing `\` before, e.g. `\:fireDamage:`.  
   The resulting message will contain the escaped emoji: `<:fireDamage:458794525050142740>`.
   
Alternatively, if you have the `server-log` enabled, you can see the id of an emoji when it's created or modified.

The following are values that always require to have an emoji assigned. These will always default to a unicode emoji.
```yaml
# Required emojis
# These values must always have an emoji to display
online_emoji: 🔹
true_emoji: ✅
false_emoji: ❌
levelup_emoji: 🌟
death_emoji: ☠
pvpdeath_emoji: 💀
# These emojis are also used in vocation filtering
novoc_emoji: 🐣
druid_emoji: ❄
sorcerer_emoji: 🔥
paladin_emoji: 🏹
knight_emoji: 🛡
# Emojis for bestiary info
charms_emoji: ⚜
difficulty_on_emoji: ⭐
difficulty_off_emoji: ▪
occurrence_on_emoji: 🔹
occurrence_off_emoji: ▪
```

The following are optional emojis. By default, these values will show text, for example creature elemental resistances.
The unicode emojis used by default are only placeholders, and it is advised to use custom emojis that better represent their values.

The `images\emoji` folder contains a set of suggested images to use for custom emojis.
```yaml
# Optional emojis
# Whether to use or not emojis to show discord status in serverinfo command.
use_status_emojis: false
status_emojis:
  online: 💚
  dnd: ♥
  idle: 💛
  offline: 🖤

# Whether to use emojis to represent elemental damage in spells and monsters info
use_elemental_emojis: false
elemental_emojis:
  physical: ⚔
  earth: 🌿
  fire: 🔥
  energy: ⚡
  ice: ❄
  death: 💀
  holy: 🔱
  poison: 🐍
```

docs/commands/settings.md
# Settings command
Commands related to server customization.

Settings can only be modified by Server Moderators (users with `Manage Server` permission).

!!! info
    Parameters are enclosed with `< >`.   
    Optional parameters are enclosed in brackets `[]`.
    
## settings
Shows the available server-specific settings that can be modified.

??? Summary "Examples"
    **/settings**  
    ![image](../assets/images/commands/settings.png)
    
----

### settings askchannel
**Syntax:** `settings askchannel [channel]`  
**Other aliases:** `settings commandchannel`

Changes the channel where longer replies for commands are given.

In this channel, pagination commands show more entries at once and command replies in general are longer.

----

### settings commandsonly
**Syntax:** `settings commandsonly [yes/no]`

Sets whether only commands are allowed in the command channel.

If this is enabled, everything that is not a message will be deleted from the command channel.
This allows the channel to be used exclusively for commands.

If the channel is shared with other command bots, this should be off.

Note that the bot needs `Manage Messages` permission to delete messages.

----

### settings eventschannel
**Syntax:** `settings eventschannel [channel]`

Changes the channel where upcoming events are announced.

This is where announcements of events about to happen will be made.
If the assigned channel is deleted or forbidden, the top channel will be used.

If this is disabled, users that subscribed to the event will still receive notifications via PM.

----

### settings levelschannel
**Syntax:** `settings levelschannel [channel]`  
**Other aliases:** `settings deathschannel`, `settings trackingchannel`

Changes the channel where levelup and deaths are announced.

This is were all level ups and deaths of registered characters will be announced.
By default, the highest channel on the list where the bot can send messages will be used.
If the assigned channel is deleted or forbidden, the top channel will be used again.

If this is disabled, Announcements won't be made, but there will still be tracking.

----

### settings minlevel
**Syntax:** `settings minlevel [channel]`  
**Other aliases:** `settings announcelevel`

Changes the channel where levelup and deaths are announced.

Level ups and deaths under the minimum level are still and can be seen by checking the character directly.

----

### settings newschannel
**Syntax:** `settings newschannel [channel]`

Changes the channel where Tibia news are announced.

This is where all news and articles posted in Tibia.com will be announced.  
If the assigned channel is deleted or forbidden, the top channel will be used.

----

### settings prefix
**Syntax:** `settings prefix [prefix]`  

Changes the command prefix for this server.

The prefix are the characters that go before a command's name, in order for the bot to recognize the command.
A maximum of 5 commands can be set per server.

To remove an existing prefix, use it as a parameter.

If you want to have a space at the end, such as: `nabbot help`, you have to use double quotes "nabbot ".
Multiple words also require using quotes.

Mentioning the bot is always a valid command and can't be changed."

----

### settings welcome
**Syntax:** `settings welcome [message]`

Changes the message new members receive when joining.
        
This is initially disabled.

You can use formatting to show dynamic values:
- {server} -> The server's name.
- {server} -> The server's owner name
- {server} -> Mention to the server's owner.
- {owner} -> The name of the server owner
- {owner.mention} -> Mention the server owner.
- {user} -> The name of the user that joined.
- {user.mention} -> Mention the user that joined.
- {bot} -> The name of the bot
- {bot.mention} -> Mention the bot.

Be sure to change the welcome channel too.

----

### settings welcomechannel
**Syntax:** `settings welcomechannel [channel]`

Changes the channel where new members are welcomed.

A welcome message must be set for this setting to work.
If the channel becomes unavailable, private messages will be used.

Note that private messages are not reliable since new users can have them disabled before joining.
To disable this, you must disable welcome messages using `settings welcome`.

----

### settings world
**Syntax:** `settings world [world]`  

Changes the world this discord server tracks.

The tracked world is the Tibia world that this discord server is following.
Only characters in that world will be registered.

----


CHANGELOG.md
# Changelog
- ✔ New feature
- 🔧 Improvement
- 🐛 Fixed bug
- ❌ Removed feature

##Version 2.1.0 (Unreleased)
- ✔ New `/announce` command for owners.
- 🐛 Fixed bug with `/makesay`
- 🐛 Fixed bug `/boss clear`
- 🐛 Fixed bugs with event editing
- 🐛 You can no longer quote messages from NSFW channels in regular ones.
- 🐛 Created watchlist channels now get proper permissions.
- 🐛 Fixed bug with `/addchar` not working with extra spaces.
- 🐛 Fixed format error in `/sql` command.

## Version 2.0.1 (2019-01-24)
- 🔧 Added `heart of destruction` as an alias for `World Devourer` in boss timers.
- 🐛 Fixed database migration importing some numeric values as strings (`announce_channel` and `announce_level`)
- 🐛 Fixed error when using `/watchlist adduser` on a user that doesn't exist.
- 🐛 Fixed bug in `/watchlist showcount` not accepting any answers.
- 🐛 Fixed incorrect hint on `/boss set`

## Version 2.0.0 (2019-01-23)
- ✔ Migrated user database from SQLite to PostgreSQL (Database migration available)
- ✔ Users can now be ignored, so the bot doesn't respond to them.
- ✔ Command usage is now saved.
- ✔ New `/commandstats` command to see command usage stats. Yes commands x4.
- ✔ Character name, world and owner history is now saved.
- ✔ Server growth stats are now saved.
- ✔ `/boss` command to set boss cooldown timers, e.g.`/boss set heart of destruction,galarzaa fidera`
- ✔ New `/channelinfo` command.
- ✔ New `/highscores global` subcommand, shows combined highscores from worlds.
- ✔ New `/checkpm` command, to check if you can receive PMs from the bot.
- ✔ New Calculators cog:
    - 🔧 Moved `/blessings`, `/stamina` and `/stats` here
    - 🔧 Improved command output of `/stats`.wa
    - ✔ `/stamina` now accepts an optional target stamina.
    - ✔ New `/stats` subcommands: `hitpoints`, `mana` and `capacity`, to calculate the minimum level needed to reach the
     target.
    - ✔ New command: `/distanceskill`, calculates the exercise weapons needed to reach a target.
    - ✔ New command: `/meleeskill`, calculates online and offline training time and exercise weapons. 
    - ✔ New command: `/magiclevel`, calculates mana needed, offline training time and exercise weapons needed.
- ✔ New Timers cog:
    - 🔧 Moved `/event` and subcommands here.
    - ✔ New `remindme` command, creates a custom reminder, e.g `/remindme 1d conquer the world`
    - ✔ New `bosstimer` command, keep track of boss cooldowns and get notified when they are over.
- 🔧 Improved and optimized TibiaWiki cog:
    - ✔ Now uses [tibiawiki-sql](https://github.com/galarzaa90/tibiawiki-sql/)'s API.
    - 🔧 Improved the display of all commands.
    - ✔ New `/charms` command
    - 🔧 `/spell` now shows the spell's effect.
    - 🔧 `/achievement` now uses new discord spoiler feature.
- 🔧 Improved event announcement task.
- 🔧 Watchlist improvements
    - 🔧 "Watched lists" are now named Watchlist.
    - ✔ You can now have multiple watchlists per server (e.g. one for friends, one for enemies).
    - ✔ New subcommand `/watchlist create`.
    - ✔ New subcommand `/watchlist adduser`, adds a user's character to a list.
    - 🔧 Improved watchlist task.
    - 🔧 Better permission management. If you have `Manage Channel` permission on the list, you can add and remove entries.
- 🔧 Server Log improvements
    - 🔧 Server log channel can no be configured (`/settings serverlog`).
    - ✔ Name changes for registered characters are now shown.
    - ✔ World transfers for registered characters are now shown.
    - 🔧 Bots get a different embed color when joining.
    - 🔧 When a member leaves or is kicked from the server, their registered characters are shown.
- 🔧 Moved `/addchar` and `/removechar` from Admin cog to Settings cog. Merged `/addaccount` and `/addchar`. 
- 🔧 `/whois` now shows Account Status, Loyalty Title and Position if any.
- 🔧 Improved performance of `/deaths`, `/levelups` and `/timeline` commands and their subcommands.
- 🔧 Death and level up tracking has been optimized, reducing unnecessary tibia.com calls and improving speed.
- 🔧 Move server timezones to their own table.
- 🔧 Improve internal logging system.
- 🔧 Created classes to handle database data.
- 🔧 Many improvements to the character tracking system.
- 🔧 Reduced news tracking interval
- 🔧 Moved `/settings` command and subcommands to Admin cog, removed Settings cog.
- 🔧 Support for multiple killers (pvp deaths).
- 🔧 Highscores task has been optimized.
- 🐛 Autorole * rule only applies to characters in the same world. This means members wont get a role for characters in a different world.
- 🐛 Bots no longer receive welcome messages.
- 🐛 `/quote` can be used on messages with only an attachment.

## Version 1.7.2 (2018-10-23)
- 🐛 Deaths caused by arena creatures are no longer announced.

## Version 1.7.1 (2018-10-12)
- 🔧 Added user caching to `/deaths`, `/levels` and `/timeline` because it was causing connection timeouts.
- 🔧 Combine walks through fields and walks around fields into a single embed field.
- 🐛 Fixed links in `/about` command.

## Version 1.7.0 (2018-09-26)
- ✔ `/monster` now shows fields monsters walk around or through
- 🔧 Improved death scanning times
- 🔧 Server admins can bypass event limit on their servers
- 🔧 Added `imbue` alias to `/imbuement` command.
- 🔧 Event announcements are now made at: 1h, 30min, 10 minutes and 0 minutes before event.

## Version 1.6.1 (2018-09-06)
- 🔧 Unified missing argument error messages
- 🔧 Commands in `/help` are now sorted alphabetically
- 🐛 Fixed bug with /worlds

## Version 1.6.0 (2018-08-27)
- ✔ New command: `/rashid`.
- ✔ Custom timezones can be added on a per-server basis using `/time add`. See `/time` subcommands for more info.
- 🔧 `/time` no longer displays Mexico and Brazil's timezones.
- 🔧 Now requires module `pytz`.
- 🔧 `/watched` can now be used by Server Moderators instead of Administrators only.
- 🐛 Fixed a bug with the global online list not having its levels updated.
- 🐛 Fixed a bug with invalid characters when using `/whois`.

cogs/utils/config.py
import os
import re
import shutil

import yaml
import yaml.reader

yaml.reader.Reader.NON_PRINTABLE = re.compile(
    u'[^\x09\x0A\x0D\x20-\x7E\x85\xA0-\uD7FF\uE000-\uFFFD\U00010000-\U0010FFFF]')

CONFIG_PATH = "config.yml"
TEMPLATE_PATH = "data/config_template.yml"

KEYS = [
    "ask_channel_name",
    "ask_channel_delete",
    "log_channel_name",
    "lite_servers",
    "owner_ids",
    "online_list_expiration",
    "announce_threshold",
    "online_scan_interval",
    "death_scan_interval",
    "network_retry_delay",
    "extra_cogs",
    "command_prefix",
    "online_emoji",
    "true_emoji",
    "false_emoji",
    "warn_emoji",
    "online_emoji",
    "true_emoji",
    "false_emoji",
    "levelup_emoji",
    "death_emoji",
    "pvpdeath_emoji",
    "novoc_emoji",
    "druid_emoji",
    "sorcerer_emoji",
    "paladin_emoji",
    "knight_emoji",
    "charms_emoji",
    "loading_emoji",
    "difficulty_on_emoji",
    "difficulty_off_emoji",
    "occurrence_on_emoji",
    "occurrence_off_emoji",
    "use_status_emojis",
    "status_emojis",
    "use_elemental_emojis",
    "elemental_emojis"
]

_DEFAULT_STATUS_EMOJIS = {
    "online": "💚",
    "dnd": "♥",
    "idle": "💛",
    "offline": "🖤"
}

_DEFAULT_ELEMENTAL_EMOJIS = {
    "physical": "⚔",
    "earth": "🌿",
    "fire": "🔥",
    "energy": "⚡",
    "ice": "❄",
    "death": "💀",
    "holy": "🔱",
    "poison": "🐍"
}


class Config:
    def __init__(self,):
        # Default values will be used if the keys are not found in config.yml
        self.ask_channel_name = "ask-nabbot"
        self.ask_channel_delete = True
        self.log_channel_name = "server_log"
        self.command_prefix = ("/", )
        self.lite_servers = []
        self.owner_ids = []
        self.extra_cogs = []
        self.online_list_expiration = 300
        self.announce_threshold = 30
        self.online_scan_interval = 90
        self.death_scan_interval = 15
        self.network_retry_delay = 1
        self.online_emoji = "🔹"
        self.true_emoji = "✅"
        self.false_emoji = "❌"
        self.warn_emoji = "⚠"
        self.levelup_emoji = "🌟"
        self.death_emoji = "☠"
        self.pvpdeath_emoji = "💀"
        self.novoc_emoji = "🐣"
        self.druid_emoji = "❄"
        self.sorcerer_emoji = "🔥"
        self.paladin_emoji = "🏹"
        self.knight_emoji = "🛡"
        self.charms_emoji = "⚜"
        self.loading_emoji = "⏳"
        self.difficulty_on_emoji = "⭐"
        self.difficulty_off_emoji = "▪"
        self.occurrence_on_emoji = "🔹"
        self.occurrence_off_emoji = "▪"
        self.use_status_emojis = False
        self.status_emojis = _DEFAULT_STATUS_EMOJIS
        self.use_elemental_emojis = False
        self.elemental_emojis = _DEFAULT_ELEMENTAL_EMOJIS

    def __repr__(self) -> str:
        kwargs = vars(self)
        attributes = []
        for k, v in kwargs.items():
            if v is None:
                continue
            if isinstance(v, int) and v == 0:
                continue
            if isinstance(v, list) and len(v) == 0:
                continue
            attributes.append(f"{k} = {v.__repr__()}")
        return f"Config({', '.join(attributes)})"

    def parse(self):
        if not os.path.isfile(CONFIG_PATH):
            print("\tconfig.yml not found, copying from template...")
            shutil.copyfile(TEMPLATE_PATH, CONFIG_PATH)
        with open(CONFIG_PATH, "r", encoding="utf-8") as f:
            _config = yaml.load(f)
            if _config is None:
                _config = {}
        self._assign_keys(_config)
        missing_keys = [k for k in KEYS if k not in _config.keys()]
        for key in missing_keys:
            print(f"\33[33m\tMissing '{key}', using default: {repr(getattr(self, key))}\033[0m")
        for key in _config:
            if key not in KEYS:
                print(f"\33[34m\tExtra entry found: '{key}'\033[0m")
        if missing_keys:
            print("\33[35mCheck data/config_template.yml for reference\033[0m")
        print("\tDone")

    def _assign_keys(self, _config):
        """Assings found keys"""
        for key in _config:
            if key == "command_prefix":
                if isinstance(_config[key], str):
                    _config[key] = (_config[key],)
                else:
                    _config[key] = tuple(_config[key])
            if key == "elemental_emojis":
                self._fill_missing_elements(key, _config[key], _DEFAULT_ELEMENTAL_EMOJIS)
            if key == "status_emojis":
                self._fill_missing_elements(key, _config[key], _DEFAULT_STATUS_EMOJIS)
            setattr(self, key, _config[key])

    @staticmethod
    def _fill_missing_elements(name, subdict, defaults):
        """Checks if the dictionary is missing keys compared against a default dictionary model."""
        missing = [k for k in defaults if k not in subdict]
        for k in missing:
            subdict[k] = defaults[k]
            print(f"\33[33m\tMissing '{k}' in '{name}', using default: {repr(defaults[k])}\033[0m")


config = Config()
